src/storage/src/memory.rs

```diff
@@ -514,6 +514,47 @@ mod batched_iter {
 
 pub type MemoryStateStore = RangeKvStateStore<BTreeMapRangeKv>;
 
+struct TableState {
+    init_epoch: u64,
+    next_epochs: BTreeMap<u64, u64>,
+    latest_sealed_epoch: Option<u64>,
+    sealing_epochs: BTreeMap<u64, BitmapBuilder>,
+}
+
+impl TableState {
+    fn new(init_epoch: u64) -> Self {
+        Self {
+            init_epoch,
+            next_epochs: Default::default(),
+            latest_sealed_epoch: None,
+            sealing_epochs: Default::default(),
+        }
+    }
+
+    async fn wait_epoch(
+        tables: &parking_lot::Mutex<HashMap<TableId, Self>>,
+        table_id: TableId,
+        epoch: u64,
+    ) {
+        loop {
+            {
+                let tables = tables.lock();
+                let table_state = tables.get(&table_id).expect("should exist");
+                assert!(epoch >= table_state.init_epoch);
+                if epoch == table_state.init_epoch {
+                    return;
+                }
+                if let Some(latest_sealed_epoch) = table_state.latest_sealed_epoch
+                    && latest_sealed_epoch >= epoch
+                {
+                    return;
+                }
+            }
+            yield_now().await;
```

Comment:
Had the code here ever caused meta CPU busy looping in some corner cases? Should we add an exponential backoff?

Reply:
The in-mem state store is used mostly for test and reference purpose. So no strong requirement for performance.

Here we assume that the all vnodes are properly assigned, so the for loop is expected to finish in short time as long as we yield the future execution here to let other parallelisms to be driven.

Reply:
Perfect. I forgot this is for the memory state store.